pkgdb/src/resolver/environment.cc

```diff
@@ -237,21 +243,22 @@ Environment::groupIsLocked( const InstallDescriptors & group,
 
 /* -------------------------------------------------------------------------- */
 
-std::vector<InstallDescriptors>
+Groups
 Environment::getUnlockedGroups( const System & system )
 {
-  auto lockfile           = this->getOldLockfile();
-  auto groupedDescriptors = this->getManifest().getGroupedDescriptors();
+  auto   lockfile           = this->getOldLockfile();
+  Groups groupedDescriptors = this->getManifest().getGroupedDescriptors();
   if ( ! lockfile.has_value() ) { return groupedDescriptors; }
 
-  for ( auto group = groupedDescriptors.begin();
-        group != groupedDescriptors.end(); )
+  for ( auto groupIterator = groupedDescriptors.begin();
+        groupIterator != groupedDescriptors.end(); )
     {
-      if ( groupIsLocked( *group, *lockfile, system ) )
+      const auto & [name, group] = *groupIterator;
+      if ( groupIsLocked( name, group, *lockfile, system ) )
         {
-          group = groupedDescriptors.erase( group );
+          groupIterator = groupedDescriptors.erase( groupIterator );
```

Comment:
caution: this smells like iterator invalidation of some sort, changing the data structure while iterating it.

resetting the pointers with `groupIterator =` and `++groupIterator` seems to prevent that, but it looks like a pitfall.

Reply:
I believe that's an acceptable pattern in C++, and it wasn't added in this PR. But I agree it's not my favorite way to write a loop